### PR TITLE
bugfix for gurobi work limit

### PR DESCRIPTION
**Pull request by tharwan**
## Fixes # 

If the WorkLimit is set in gurobi and the optimisation is stoped because the limit is reached pyomo does not detect the feasible solution.

## Summary/Motivation:

Gurobi 9.5 introduced a deterministic [work limit](https://www.gurobi.com/documentation/9.5/refman/worklimit.html) as alternative to a time limit. Pyomo does not detect this correctly if the optimisation is stopped because of WORK_LIMIT.

Work limit would be very useful for us and we would love to see a quick bugfix release including this fix.

## Changes proposed in this PR:
- handle a stop due to work limit the same as time limit

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/plugins/solvers/GUROBI_RUN.py [hunk 1]
```diff
@@ -167,6 +167,11 @@ def gurobi_run(model_file, warmstart_file, soln_file, mipgap, options, suffixes)
         message = 'Optimization terminated because the time expended exceeded the value specified in the TimeLimit parameter.'
         term_cond = 'maxTimeLimit'
         solution_status = 'stoppedByLimit'
+    elif hasattr(GRB, "WORK_LIMIT") and (solver_status == GRB.WORK_LIMIT):
+        status = 'aborted'
+        message = 'Optimization terminated because the work expended exceeded the value specified in the WorkLimit parameter.'
+        term_cond = 'maxTimeLimit'
```
Comment (michaelbynum) [inline pyomo/solvers/plugins/solvers/GUROBI_RUN.py:173]: Would `'maxIterations'` or `'maxEvaluations'` be more appropriate than `'maxTimeLimit'`?

**Review by michaelbynum — CHANGES_REQUESTED** [review]
@tharwan, thanks for this contribution. Can you please add a test for this functionality?

**Comment by tharwan** [comment]
> @tharwan, thanks for this contribution. Can you please add a test for this functionality?

I can try. Any hints where to start?

File: pyomo/solvers/plugins/solvers/GUROBI_RUN.py (hunk 1 above)
Comment (tharwan) [reply pyomo/solvers/plugins/solvers/GUROBI_RUN.py:173]: I thought about this but in the docs it is an alternative to a time limit. 🤷‍♂️

Comment (michaelbynum) [reply pyomo/solvers/plugins/solvers/GUROBI_RUN.py:173]: Good enough for me.

**Comment by tharwan** [comment]
@michaelbynum I really need some help where to even find the tests that are responsible for this part.

**Comment by michaelbynum** [comment]
I would recommend putting them in pyomo/solvers/tests/checks. I don't see any gurobi-specific tests in there right now, so I would create a new file (like we have for gurobi persistent).

**Comment by tharwan** [comment]
I don't think this is very helpful to be honest, but here you go.

**Comment by tharwan** [comment]
@michaelbynum any suggestions why the tests failed 3.7? Can we rerun the pipeline?

File: pyomo/solvers/tests/checks/test_gurobi.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import unittest
```
Comment (mrmundt) [inline pyomo/solvers/tests/checks/test_gurobi.py:1]: Pyomo-specific suggestion: we have an extended implementation of `unittest`, so we prefer:
```
import pyomo.common.unittest as unittest
```

It still has all of the standard `unittest` libraries - just adds some extra functionality for some of our testing choices.

File: pyomo/solvers/tests/checks/test_gurobi.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+try:
+    from pyomo.solvers.plugins.solvers.GUROBI_RUN import gurobi_run
+    from gurobipy import GRB
+    gurobipy_available = True
+except:
+    gurobipy_available = False
+
+class GurobiTest(unittest.TestCase):
+
+    @unittest.skipIf(not gurobipy_available,
```
Comment (mrmundt) [inline pyomo/solvers/tests/checks/test_gurobi.py:13]: Suggestion: move this first `skipIf` to right before the class declaration. That should fix the failure on the `slim` build.

**Comment by tharwan** [comment]
this is quite frustrating.

**Comment by mrmundt** [comment]
@tharwan  - I totally understand. I have some time in a couple hours to chase down some of your bugs.

**Comment by mrmundt** [comment]
@tharwan - Just pending this very last test (fingers crossed)!

**Review by mrmundt — APPROVED** [review]

**Pull request merged**